File: backend/app/schemas/evidence.py
import uuid
from datetime import datetime
from typing import Annotated

from pydantic import ConfigDict, Field, StringConstraints

from app.domain.enums import EvidenceType
from app.schemas.case import PaginationMeta
from app.schemas.common import ReadSchema, TimezoneAwareSchema


class EvidenceRegistration(TimezoneAwareSchema):
    model_config = ConfigDict(extra="forbid")

    description: str | None = None
    evidence_type: EvidenceType
    collected_by: str | None = Field(default=None, max_length=255)
    collected_at: datetime | None = None
    custody_person: Annotated[
        str, StringConstraints(strip_whitespace=True, min_length=1, max_length=255)
    ]
    custody_location: str | None = Field(default=None, max_length=500)
    custody_notes: str | None = None


class EvidenceCreate(TimezoneAwareSchema):
    """Internal Phase 1 metadata schema; not accepted by the upload API."""

    evidence_number: str = Field(min_length=1, max_length=100)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    evidence_type: EvidenceType
    original_path: str | None = None
    stored_path: str | None = None
    size_bytes: int | None = Field(default=None, ge=0)
    collected_by: str | None = Field(default=None, max_length=255)
    collected_at: datetime | None = None


class EvidenceUpdate(TimezoneAwareSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    evidence_type: EvidenceType | None = None
    original_path: str | None = None
    stored_path: str | None = None
    size_bytes: int | None = Field(default=None, ge=0)
    collected_by: str | None = Field(default=None, max_length=255)
    collected_at: datetime | None = None


class EvidenceRead(ReadSchema):
    id: uuid.UUID
    case_id: uuid.UUID
    evidence_number: str
    name: str
    description: str | None
    evidence_type: EvidenceType
    size_bytes: int | None
    collected_by: str | None
    collected_at: datetime | None
    created_at: datetime
    updated_at: datetime


class EvidenceHashRead(ReadSchema):
    id: uuid.UUID
    evidence_id: uuid.UUID
    algorithm: str
    digest: str
    computed_at: datetime
    purpose: str


class EvidenceListResponse(ReadSchema):
    data: list[EvidenceRead]
    meta: "PaginationMeta"


class EvidenceVerificationRead(ReadSchema):
    evidence_id: uuid.UUID
    algorithm: str
    recorded_digest: str
    calculated_digest: str
    match: bool
    verified_at: datetime

File: frontend/src/types/evidence.ts
import type { PaginationMeta } from './case'

export type EvidenceType = 'DISK_IMAGE' | 'LOG_FILE' | 'MEMORY_DUMP' | 'REGISTRY_HIVE' | 'FILE' | 'DIRECTORY' | 'OTHER'

export interface EvidenceRecord {
  id: string
  case_id: string
  evidence_number: string
  name: string
  description: string | null
  evidence_type: EvidenceType
  size_bytes: number | null
  collected_by: string | null
  collected_at: string | null
  created_at: string
  updated_at: string
}

export interface EvidenceListResponse { data: EvidenceRecord[]; meta: PaginationMeta }
export interface EvidenceHashRecord { id: string; evidence_id: string; algorithm: string; digest: string; computed_at: string; purpose: string }
export interface CustodyEntry { id: string; evidence_id: string; timestamp: string; person: string; action: string; location: string | null; notes: string | null; created_at: string }
export interface EvidenceVerification { evidence_id: string; algorithm: string; recorded_digest: string; calculated_digest: string; match: boolean; verified_at: string }
export interface EvidenceRegistrationRequest {
  file: File
  evidence_type: EvidenceType
  custody_person: string
  description?: string
  collected_by?: string
  collected_at?: string
  custody_location?: string
  custody_notes?: string
}
